lib/ringbuffer_helper.c

```diff
@@ -306,3 +306,86 @@ qb_rb_sem_create(struct qb_ringbuffer_s * rb, uint32_t flags)
 	}
 	return rc;
 }
+
+int32_t
+qb_rb_close_helper(struct qb_ringbuffer_s * rb, int32_t unlink_it,
+		   int32_t truncate_fallback)
+{
+	int32_t res = 0, res2 = 0;
+	uint32_t word_size = rb->shared_hdr->word_size;
+	char *hdr_path = rb->shared_hdr->hdr_path;
+
+	if (unlink_it) {
+		qb_util_log(LOG_DEBUG, "Free'ing ringbuffer: %s", hdr_path);
+		if (rb->notifier.destroy_fn) {
+			(void)rb->notifier.destroy_fn(rb->notifier.instance);
+		}
+	} else {
+		qb_util_log(LOG_DEBUG, "Closing ringbuffer: %s", hdr_path);
+		hdr_path = NULL;
+	}
+
+	if (unlink_it) {
+#if defined(HAVE_OPENAT) && defined(HAVE_UNLINKAT)
+#  ifndef O_DIRECTORY
+#    define RB_DIR_RO_FLAGS1 O_RDONLY
+#  else
+#    define RB_DIR_RO_FLAGS1 O_RDONLY|O_DIRECTORY
+#  endif
+#  ifndef O_PATH
+#    define RB_DIR_RO_FLAGS RB_DIR_RO_FLAGS1
+#  else
+#    define RB_DIR_RO_FLAGS RB_DIR_RO_FLAGS1|O_PATH
+#  endif
+		char *data_path = rb->shared_hdr->data_path;
+		char *sep = strrchr(data_path, '/');
```

Comment:
See https://github.com/ClusterLabs/libqb/pull/231/commits/15591922346ad485eb054f4c611b226dc6315393#diff-55541c07268e810475b8e7d661314605R311
Thanks for making me review the actual commits -- I've discovered that I had accidentally squashed two commits into what was the last one.  Now they are split again.